### ISSUE TITLE
Incredible slow mathbb rendering

### ISSUE THREAD

**Issue by memeplex** [mathjax/MathJax#2132]
A page with simply `{\mathbb{R}}` as its full content takes about 10 seconds (!!!) to render.

This happens not only the first time I open the page but after each reloading, it doesn't seem to be a matter of font caching but I can't say for sure.

I'm using the the TeX-AMS_HTML config.

**Comment by memeplex** [comment]
Indeed everything seems to be cached the second time I load the page:

![image](https://user-images.githubusercontent.com/2845433/54497148-0ffdc000-48d6-11e9-8daa-0912389da760.png)

Anyway I have to wait about 10 seconds to get the R correctly rendered. Is that normal? Something like $x^2$ renders instantaneously, everything else being the same.

**Comment by memeplex** [comment]
Another fact: the last two lines in the screenshot above, while they say "cached", appear once the final render is done, that is about 10 seconds later than the rest of the above lines.

**Comment by pkra** [comment]
This sounds like the standard 15s timeout with a bad loadComplete call. But without a live page, its hard to tell.

**Comment by memeplex** [comment]
Could you check how much this take to produce a correctly rendered font for you?

```html
<html>
<head>
<script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS_HTML"></script>
</head>
<body>
<div style="display:none">
\[ \newcommand{\R}{\mathbb{R}} \]
</div>
\( \R \)
</body>
</html>
```

Thanks.

**Comment by pkra** [comment]
Around 1.71s from a clean browser cache, about 0.2s between the last part of MathJax's parts arriving and the final rendering.

**Comment by dpvc** [comment]
The page loads quickly for me as well.  Peter is right that this sounds like the 10 second timeout that MathJax uses while waiting for a file that is not loading.

Try entering `MathJax.Message.Log()` in the browser's console window and see if there is anything there about files not being able to be loaded.

**Comment by memeplex** [comment]
There seems to be a problem loading fonts, indeed.

```
MathJax.Message.Log()

Loading [MathJax]/config/TeX-AMS_HTML.js
Loading [MathJax]/extensions/TeX/newcommand.js
Loading [MathJax]/jax/output/HTML-CSS/jax.js
Loading [MathJax]/jax/output/HTML-CSS/fonts/TeX/fontdata.js
Loading [MathJax]/jax/output/HTML-CSS/fonts/TeX/AMS/Regular/Main.js
Loading [MathJax]/jax/output/HTML-CSS/fonts/TeX/AMS/Regular/BBBold.js
Loading web-font TeX/AMS/Regular
Can't load web font TeX/AMS/Regular
Firefox can't load web fonts from a remote host
Web-Fonts not available -- using image fonts instead
Loading [MathJax]/jax/output/HTML-CSS/imageFonts.js
Loading [MathJax]/fonts/HTML-CSS/TeX/png/imagedata.js"
```

**Comment by pkra** [comment]
Sounds like an installation issue. Are the fonts at those locations? If they are not on the same domain, have you set access control headers for the fonts? Cf. https://docs.mathjax.org/en/latest/installation.html#notes-about-shared-installations

**Comment by memeplex** [comment]
Ok, thanks. I don't think this is my case:

> There is a solution to this, however, if you manage the server where MathJax is installed, and if that server is running the Apache web server. 

Short of this, can I reduce the timeout at least?

OTOH I assume this is working for me despite not being able to download web fonts because I've the right fonts locally installed. Now, supposing that anyone else is accessing my page from firefox and not being able to download the web fonts, what would be the consequence? No latex rendering at all? Or just some broken parts  here and there?

**Comment by dpvc** [comment]
There are a number of reasons why MathJax might be having trouble with the fonts.  One is Peter's suggestion that you have an installation issue.  Are you using MathJax from a CDN, or from your own server?

Another is that the fonts actually are getting through, but that MathJax is having trouble detecting that.  The mechanism used to detect when a web font has arrived is rather fragile, and can be affected by the CSS on your page.  Did the test page that you listed above also show the delay?

A third possibility is that you have disabled page-specified font in your browser, which would prevent MathJax from changing to the web fonts at all, and that would certainly cause the problem you are having.

> I assume this is working for me despite not being able to download web fonts because I've the right fonts locally installed.

No, MathJax is not using locally installed fonts for you because it is trying to load the web-based fonts.  If it had detected the local fonts, it would not be trying to load web versions. The reason it works for you is because it has switched to images rather than true fonts as a fallback to get you something that works.  You can see this in the last lines of the message log:

```
Web-Fonts not available -- using image fonts instead
Loading [MathJax]/jax/output/HTML-CSS/imageFonts.js
Loading [MathJax]/fonts/HTML-CSS/TeX/png/imagedata.js"
```

So the result for others using your site when this happens to them should be the same result as you are seeing.

> Short of this, can I reduce the timeout at least?

You can, but I wouldn't recommend it.  It would be better to figure out why the fonts are not arriving (or why they aren't being detected).  Also, if you set it too short, you might end up incorrectly thinking the font isn't arriving when you actually haven't waited long enough.  The default is 8 seconds for a desktop computer and 15 for a mobile device (since they tend to have slower connections).  If you must, you could change these using something like

```
<script type="text/x-mathjax-config">
MathJax.Hub.Register.StartupHook("HTML-CSS Jax Startup") {
    var HTMLCSS = MathJax.OutputJax["HTML-CSS"];
    HTMLCSS.Font.timout = (MathJax.Hub.Browser.isMobile ? 10 : 4) * 1000;
}
</script>
```
just **before** the script that loads MathJax.js itself.  This should reduce the delay to 10 seconds for mobile and 4 for desktop, but I wouldn't go any lower than that.

**Comment by memeplex** [comment]
Ok, thanks again.

> It would be better to figure out why the fonts are not arriving (or why they aren't being detected).

Fine. I'm answering your questions below. Please let me know if you need additional debugging from my side.

> Are you using MathJax from a CDN, or from your own server?

Just the snippet I shared above. Besides cloudflare I tested rawgit but the results were the same.

> Did the test page that you listed above also show the delay?

Yes, the delay was reported for that page, the reported logs are also from loading that page from my local file system.

> A third possibility is that you have disabled page-specified font in your browser

Doesn't seem to be the case:

![image](https://user-images.githubusercontent.com/2845433/54842126-ce8f5b00-4caf-11e9-892d-bee601e73695.png)

**Comment by dpvc** [comment]
Are there any plugins the might be interfering with the fonts?  Have you tried starting Firefox without plugins (see [these instructions](https://support.mozilla.org/en-US/kb/troubleshoot-extensions-themes-to-fix-problems))?

**Comment by memeplex** [comment]
Still getting the same noticeable delay with no addon enabled.

**Comment by dpvc** [comment]
Do you have a minimum font size set?

I'm running out of ideas for what might be the problem.   It might be that the CDN edge-node that is closest to you is having a problem and isn't serving the font, but the fact that you had the same problem with rawgit would tend to preclude that.  I suppose there might be some issue with a firewall or router that is configured to not forward the fonts; do you see web fonts on other pages?

My only other suggestion would be to switch from the HTML-CSS output to CommonHTML or SVG output (neither of which have to wait for fonts, and so have no font delays).  Use `TeX-AMS_CHTML` or `TeX-AMS_SVG` instead of `TeX-AMS_HTML`.

**Comment by memeplex** [comment]
> I suppose there might be some issue with a firewall or router

I don't think so but anyway I'll try connecting from another place later and report the outcome here.

> Use TeX-AMS_CHTML or TeX-AMS_SVG instead of TeX-AMS_HTML

Both of these worked as a charmed. I don't even know what Common HTML is, will have to research this. Is there any drawback with these backends that I should worry about?

**Comment by memeplex** [comment]
Ok, this settles the question for me

> The CommonHTML output processor produces high-quality output in all modern browsers, with results that are consistent across browsers and operating systems. This is MathJax’s primary output mode since MathJax v2.6. Its major advantage is its quality, consistency, and speed as well as support for server-side generation.

If you are still interested in finding the underlying issue I've no problem in further debugging. But from a practical perspective I'm ok with CHTML, I was just using what I used for a long time without thinking about it.

**Comment by memeplex** [comment]
What is strange though is that both CHTML and HTML-CSS are announced as

>  uses web-based fonts so that users don’t have to...

but, nevertheless, the delay is not happening here for CHTML. That discards the firewall hypothesis, I guess.

**Comment by dpvc** [comment]
Yes, both HTML-CSS and CommonHTML use web fonts.  But CommonHTML doesn't try to detect their presence, while HTML-CSS does, and it is that detection that is probably going wrong for some reason (I did say it was fragile).  HTML-CSS needed to wait for the fonts to arrive because (in the old days before browsers were more consistent with each other), MathJax had to measure the actual displayed subexpressions in order to determine their size, and that meant the font had to be the right one.  CommonHTML doesn't have to do that because modern browsers are more consistent about their layout (thanks to better standards and better adherence to them).

One last thing to check:  did you change the default font in your browser, or have you set a default stylesheet that is not the standard one?  It is possible this could cause the detection to fail, though it is unlikely you have changed either of those.

**Comment by dpvc** [comment]
> That discards the firewall hypothesis, I guess.

Yes, that does seem to be the case.  It was a longshot idea in any case.

**Comment by memeplex** [comment]
>  did you change the default font in your browser

Some of them, but still a rather standard setting:

![image](https://user-images.githubusercontent.com/2845433/54845448-1a460280-4cb8-11e9-83ae-117c50dd96d7.png)

> have you set a default stylesheet that is not the standard one?

No, I haven't. I try not to do fancy things with my browser.

**Comment by dpvc** [comment]
OK, I don't think those fonts should be any problem.  I'm stumped.  But at least you can use CommonHTML instead.

**Comment by memeplex** [comment]
Yes, it's ok for me. In fact, I'm better than before in that I learned I was using an somewhat outdated backend. Close this if you want. Thank you very much!

**Comment by dpvc** [comment]
I think you'll find that CommonHTML is faster at rendering the math, so you should like that.

Good luck!
